source/slang/slang-capability.h

```diff
@@ -46,108 +48,52 @@ namespace Slang
-    //
-    // Note that we are using the term "incompatible" here even though it
-    // seems like "disjoint" would be intuitively correct (HLSL and GLSL
-    // targets sure do seem to be disjoint). The problem is that in our
-    // set-theoretic representation of capabilities, incompatible capability
-    // sets are *never* disjoint sets of atoms, and (valid) disjoint sets of atoms
-    // *never* represent incompatible capability sets.
-
-        /// Is this capability set incompatible with the given `other` set.
-    bool isIncompatibleWith(CapabilityAtom other) const;
 
-        /// Is this capability set incompatible with the given `other` atomic capability.
-    bool isIncompatibleWith(CapabilityConjunctionSet const& other) const;
-
-    // One capability set A "implies" another set B if a target that
-    // supports A must also support all of B.
-    //
-    // In practice, this means that "A implies B" is the same as
-    // "A is a subset of B" in the set-theoretic model, but
-    // we ant to think of this primarily as supported/required features,
-    // and not get hung up on the set theory.
-
-        /// Does this capability set imply all the capabilities in `other`?
-    bool implies(CapabilityConjunctionSet const& other) const;
-
-
-        /// Does this capability set imply the atomic capability `other`?
-    bool implies(CapabilityAtom other) const;
-
-    // A capability set is equal to another if each implies the other.
-
-        /// Are these two capability sets equal?
-    bool operator==(CapabilityConjunctionSet const& that) const;
-    bool operator<(CapabilityConjunctionSet const& that) const;
-
-        /// Get access to the raw atomic capabilities that define this set.
-    List<CapabilityAtom> const& getExpandedAtoms() const { return m_expandedAtoms; }
-    List<CapabilityAtom>& getExpandedAtoms() { return m_expandedAtoms; }
-
-        /// Calculate a list of "compacted" atoms, which excludes any atoms from the expanded list that are implies by another item in the list.
-    void calcCompactedAtoms(List<CapabilityAtom>& outAtoms) const;
+struct CapabilityAtomSet : UIntSet 
+{
+    using UIntSet::UIntSet;
+};
 
-    Int countIntersectionWith(CapabilityConjunctionSet const& that) const;
+struct CapabilityTargetSet;
+typedef Dictionary<CapabilityAtom, CapabilityTargetSet> CapabilityTargetSets;
 
-    bool isBetterForTarget(CapabilityConjunctionSet const& that, CapabilityConjunctionSet const& targetCaps) const;
+/// CapabilityStageSet encapsulates all capabilities of a specific shader stage for a specific target.
+/// Capabilities may be disjoint, but only in rare cases: 
+/// {{glsl, _GLSL_130, GL_EXT_FOO1}, {glsl, _GLSL_130, _GLSL_140, _GLSL_150}}
+struct CapabilityStageSet
+{
+    CapabilityAtom stage{};
+
+    /// LinkedList of all disjoint sets for fast remove/add of unconstrained list positions.  
+    std::optional<CapabilityAtomSet> atomSet{};
+
+    void addNewSet(const CapabilityAtomSet& setToAdd)
```

Comment:
Prefer just pass CapabilityAtomSet by value to enable copy ellision. Can provide an CapabilityAtomSet&& for guaranteed move semantics.